doc/sphinx/proofs/writing-proofs/reasoning-inductives.rst

```diff
@@ -5,269 +5,297 @@ Reasoning with inductive types
-      chaining destruction of a hypothesis.
+The tactics in this section implement case
+analysis on inductive or co-inductive objects (see :ref:`gallina-inductive-definitions`).
 
-   .. tacv:: destruct @term eqn:@naming_intropattern
-      :name: destruct … eqn:
+Some useful terms:
 
-      This behaves as :n:`destruct @term` but adds an equation
-      between :token:`term` and the value that it takes in each of the
-      possible cases. The name of the equation is specified
-      by :token:`naming_intropattern` (see :tacn:`intros`),
-      in particular ``?`` can be used to let Coq generate a fresh name.
+* `n` is a :gdef:`dependent premise` of `forall n:nat, n + 0 = n` because
+  `n` appears both in the binder of the `forall` and in the quantified statement
+  `n + 0 = n`.
 
-   .. tacv:: destruct @term with @bindings
+* `A` and `B` are :gdef:`non-dependent premises <non-dependent premise>` of `A -> B -> C`
+  because they don't appear in a `forall` binder.
+  (Recall that `A -> B` is notation for the term `forall _ : A, B`; the wildcard
+  `_` can't be referred to in the quantified statement.)
 
-      This behaves like :n:`destruct @term` providing explicit instances for
-      the dependent premises of the type of :token:`term`.
+.. tacn:: destruct @induction_clause_list
 
-   .. tacv:: edestruct @term
-      :name: edestruct
+   .. insertprodn induction_clause_list as_or_and_ipat
 
-      This tactic behaves like :n:`destruct @term` except that it does not
-      fail if the instance of a dependent premises of the type
-      of :token:`term` is not inferable. Instead, the unresolved instances
-      are left as existential variables to be inferred later, in the same way
-      as :tacn:`eapply` does.
+   .. prodn::
+      induction_clause_list ::= {+, @induction_clause } {? using @one_term {? with @bindings } } {? @occurrences }
+      induction_clause ::= @destruction_arg {? @as_or_and_ipat } {? eqn : @naming_intropattern } {? @occurrences }
+      destruction_arg ::= @natural
+      | @one_term_with_bindings
+      as_or_and_ipat ::= as {| @or_and_intropattern | @ident }
+      | as @equality_intropattern
 
-   .. tacv:: destruct @term using @term {? with @bindings }
+   Generates a subgoal for each constructor of the inductive or co-inductive
+   type in the :n:`@one_term` in :n:`@destruction_arg`.  Unlike :tacn:`induction`,
+   no induction hypothesis is generated by :tacn:`destruct`.
 
-      This is synonym of :n:`induction @term using @term {? with @bindings }`.
+   The :n:`@one_term` may contain holes that are denoted by “_”. In this case,
+   the tactic checks that all subterms matching the pattern in the conclusion
+   and the hypotheses are compatible and performs case analysis using this subterm.
 
-   .. tacv:: destruct @term in @goal_occurrences
+   .. todo PR Need more explanation about the tactic does when destruction_arg is
+      a one_term that's not just an identifier'
```

Comment:
Example from a recent proof I did:

```coq
destruct (compare a1 b1) eqn:Hab
```

Does a case analysis on `compare a1 b1` (can be `Eq`, `Lt` or `Gt`). For each case, thanks to the `eqn:` clause, it will generate a new hypothesis of the form `Hab : compare a1 b1 = Eq (* or Lt or Gt *)`. In this case, the `one_term` doesn't even need to appear in the goal or the hypotheses.

Reply:
Interesting.  I didn't realize that `eqn:` would create additional hypotheses.

Without eqn, the behavior is to create similar equations for an inductive-typed value and then substitute the rhs wherever the lhs appeared in the context?

Separately, what exactly is ` { ... } + { ... }`?  It also seems odd that Coq picks `e` and `n0` for the introduced hypotheses rather than something with a capital letter.

1 subgoal
n : nat
IHn : forall m0 : nat, {n = m0} + {n <> m0}
m : nat
______________________________________(1/1)
{S n = S m} + {S n <> S m}

1: {
 
![image](https://user-images.githubusercontent.com/1253341/122143083-8f580200-ce05-11eb-9ead-eeb289455b84.png)

2 : {
 
![image](https://user-images.githubusercontent.com/1253341/122143092-954de300-ce05-11eb-867d-65d3d02adf04.png)

Reply:
> Separately, what exactly is { ... } + { ... }?

A notation for Coq.Init.Datatypes.sumbool, I believe.  (Might be in Logic or Specif rather than Datatype.). Note that you can answer most of these questions yourself by doing Locate or About on the notation string and then Locate again on the identifier (so `Locate "{ _ } + { _ }". Locate sumbool.` in this case)

Reply:
Or just `About "{ _ } + { _ }".`

>Without eqn, the behavior is to create similar equations for an inductive-typed value and then substitute the rhs wherever the lhs appeared in the context?

Yes. Note that with `eqn`, the substitution also occurs. But it's still useful to get the equation for later.

>  It also seems odd that Coq picks `e` and `n0` for the introduced hypotheses rather than something with a capital letter.

Yes, I don't know why it is so.

Reply:
> > It also seems odd that Coq picks `e` and `n0` for the introduced hypotheses rather than something with a capital letter.
> 
> Yes, I don't know why it is so.

`e` from the first letter of the type `eq`, `n0` from the first letter of the type `not`, I think

Reply:
The curly braces mean that it's a set?

Further off track: FWIW, I don't understand what `Specify.v` is useful for.  Some of the comments make me think it's partly about manipulating dependent types--something I tried to figure out a long while ago but didn't get anywhere.  (Not the concepts, but the nuts and bolts of using them.)

For example:

```
(** Projections of [sig]

    An element [y] of a subset [{x:A | (P x)}] is the pair of an [a]
    of type [A] and of a proof [h] that [a] satisfies [P].  Then
    [(proj1_sig y)] is the witness [a] and [(proj2_sig y)] is the
    proof of [(P a)] *)
```

Reply:
> The curly braces mean that it's a set?

Curly braces seem to indicate turning a `Prop` into a `Set`
https://github.com/coq/coq/blob/147f4da1beee028be550f098401398f5822af78e/theories/Init/Specif.v#L887-L890



> Further off track: FWIW, I don't understand what `Specify.v` is useful for.

I believe `Specif.v` is short for "specifications" (which indeed are largely based on using dependent types)

Reply:
Thanks.  I expect the standard library is not well documented in general.  Something to look at later.